### ISSUE TITLE
Possible leak in ctx_init_libp11_unlocked()

### ISSUE THREAD

**Issue by kristjan79** [OpenSC/libp11#228]
When the `PKCS11_CTX_load()` fails because of errors from the `C_Initialize()` or `C_GetInfo()` calls then the module will not be unloaded since only `PKCS11_CTX_free()` is called and hence it leaks.

A rough fix would be to always call `PKCS11_CTX_unload()` also if `PKCS11_CTX_load()` fails, though in that case the `C_UnloadModule()` would return an error if the shared library could not be loaded (this would be currently ignored). No doubt, there are more intelligent ways to handle this as well.

However in the latest code the module is not unloaded on `ctx_finish()` call at all, so I do not know how much of a problem it is considered to be. A few remarks about this as well:
- It is not just about memory leak vs deadlock. There are also dynamic library constructor and destructor functions that need to be considered. If the module leaks then the destructor will no longer be called as reference count will never reach 0 and also on the next init it will no longer call the contructor since the module is already mapped. If these routines are used by the library provider then the calls might be vital in order to get out of an error situation. Though in general I agree that deadlock would be even worse.

- Wouldn't it still make sense to call `C_Finalize()` event if the module cannot be unloaded? Currently it seems that `C_Finalize()` is only called from `pkcs11_CTX_unload()` so I guess it is not called.

**Comment by frankmorgner** [comment]
I think the error handling in `pkcs11_CTX_load` should be improved. If it errors during `C_Initialize` or `C_GetInfo` it needs to rollback all the initialized data.

**Comment by frankmorgner** [comment]
please check whether https://github.com/OpenSC/libp11/pull/229 fixes the problem

**Comment by kristjan79** [comment]
Sorry for the delay. Yes on version 0.4.2 what we currently use for RHEL7 it would solve the whole problem, thank You. 

Could You please comment on `C_Finalize` as well, is it handled as it is supposed? 
Under normal flow it will currently never be called since it is currently only called together with unload and `ctx_finish` at this point does not call unload. Meaning if the module was loaded then it will never be unloaded and it will also never call `C_Finalize`.

**Comment by frankmorgner** [comment]
- `PKCS11_CTX_load` will call `C_Finalize` itself if it receives an error after `C_Initialize`.
- An application that calls `PKCS11_CTX_load` without receiving an error should eventually call `PKCS11_CTX_unload`?

This means that a successful call to `C_Initialize` is always followed by `C_Finalize`.

**Comment by kristjan79** [comment]
I can see `PKCS11_CTX_unload` called by two functions. 
- `ctx_init_libp11_unlocked` in case something goes wrong during load.
- `ctx_finish` when the engine is destructed, but this one is currently commented out. 
```
/* Modules cannot be unloaded in pkcs11_finish() nor
 * ctx_destroy() because of a deadlock in PKCS#11
 * modules that internally use OpenSSL engines.
 * A memory leak is better than a deadlock... */
/* PKCS11_CTX_unload(ctx->pkcs11_ctx); */
```
Maybe I am missing something, but it seems to me that once the module is loaded. The only way to unload it would be to terminate the process and the OS will take care of this. This also implies that the `C_Finalize` will not be called.

**Comment by kristjan79** [comment]
Sorry I probably should have created two separate issues for these

**Comment by frankmorgner** [comment]
I'm not familiar with the ENGINE-design, maybe  @mtrojnar can check this problem.

**Comment by mtrojnar** [comment]
Could you clearly define *this problem* please?  This thread seems to mention a few separate problems...
What is the expected behavior?  What is the actual behavior?
This may help: https://www.chiark.greenend.org.uk/~sgtatham/bugs.html

**Comment by kristjan79** [comment]
Will try to be short and to the point (I am afraid it is still going to be long and confusing). 

**Set-up:** Our server SW uses Safenet network HSM and it is used through the OpenSSL "pkcs11" ENGINE with a proprietary module provided by Safenet (libcryptoki.so).

**Problem:** On some very rare occasions it has been noticed that the server SW loses connection to the network HSM and is unable to recover from the error situation. Usually when any kind of problem occurs it is enough to release/reset the OpenSSL ENGINE and initialize it again and the communication will be restored. In these particular situations this does not work, the only way to recover is to restart the SW process.

On closer inspection it was identified that the problem occurred, because a network switch had rebooted left the TCP session with the HSM module in some kind of limbo state. So far it seems that the only way to recover is to make sure that the Saefenet module is properly unloaded. Because the module is proprietary it is pretty much a black box. However, it seems that the Safenet driver does some kind of clean-up in the shared library destructor/_fini and hence when the module is properly unloaded it will always recover.

We can replicate the error situation in lab conditions.
- In the version libp11 version 0.4.2 (version used in the live set-up) if the SW tried to recover the ENGINE before the network had recovered it failed during `C_Initialize` call in `PKCS11_CTX_load` and after this since the module descriptor leaked it was not able to recover until the process was restarted. This is now solved with the fix made by @frankmorgner.
- In the latest master code however it is still not able to recover from the error situation, since the module is not unloaded when ENGINE is released. The only way to recover is to restart the daemon process.

In short:
**Expected result:** When the OpenSSL ENGINE is finished and released the it should call `C_Finalize` and also unload the module (shared library).
**Actual result:** Neither is done, because the code that is responsible for this in `ctx_finish` is commented out.
```
/* Modules cannot be unloaded in pkcs11_finish() nor
 * ctx_destroy() because of a deadlock in PKCS#11
 * modules that internally use OpenSSL engines.
 * A memory leak is better than a deadlock... */
/* PKCS11_CTX_unload(ctx->pkcs11_ctx); */
```

**Comment by mtrojnar** [comment]
Thank you.  The related issue is https://github.com/OpenSC/libp11/issues/75.
I have some ideas for fixing it.  Apparently `PKCS11_CTX_unload(ctx->pkcs11_ctx);` can be safely uncommented with OpenSSL 1.1.0 and later.  I will test it and submit a fix.

**Comment by mtrojnar** [comment]
@kristjan79 Please let me know if my solution works for you.

**Comment by kristjan79** [comment]
Yes, this works (with OpenSSL >= 1.1.0)
